File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, nums: List[int]) -> int:
        
        nse = [len(nums)-i for i in range(len(nums))]
        stack = []
        for i in range(len(nums)):
            while stack and nums[stack[-1]]>nums[i]:
                x = stack.pop()
                nse[x] = i-x
            stack.append(i)
        
        pse = [len(nums) - i for i in range(len(nums)-1,-1,-1)]
        stack = []
        for i in range(len(nums)-1,-1,-1):
            while stack and nums[stack[-1]]>=nums[i]:
                x = stack.pop()
                pse[x] = x - i
            stack.append(i)
        print(nse)
        print(pse)
        ans =0

        for i in range(len(nums)):
            ans += (nums[i]*nse[i]*pse[i])%1000000007
        return ans%1000000007

File: 943-sum-of-subarray-minimums/sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] l) {
        long res =0;
        int mod = 1000000007;
        Stack<Integer> s = new Stack<>();
        int[] arr = new int[l.length+1];
        for(int i =0;i<l.length;i++) arr[i] = l[i];
        arr[l.length] = Integer.MIN_VALUE;
        // arr.add(Integer.MIN_VALUE);
        for(int i =0;i<arr.length;i++){
            while(s.size()>0 && arr[s.peek()]>arr[i]){
                int q = s.pop();
                int lft = q - ((s.size()>0)? s.peek() : -1);
                int rgt = i - q;
                res += ((long)arr[q]*lft*rgt)%mod;
                res %= mod;
            }
            s.add(i);
        }
        res %=mod;
        return (int)res;
    }
}